File: 2022/adventofcode/day04.py
import functools

from aoc import load_day


__day__ = 4

def part1(data):
    def intervals_overlap(first, second) -> bool:
        return (
            (first[0] <= second[0] and second[1] <= first[1])
            or (second[0] <= first[0] and first[1] <= second[1])
        )
    return sum(1 for first, second in data if intervals_overlap(first, second))


def part2(data):
    def intervals_overlap(first, second) -> bool:
        return bool(set(range(first[0], first[1] + 1)).intersection(set(range(second[0], second[1] + 1))))
    
    return sum(1 for first, second in data if intervals_overlap(first, second))

def parse_input(data):
    def parse_row(row: str):
        return tuple(tuple(map(int, token.split('-'))) for token in row.split(','))

    return list(parse_row(row) for row in data)

data = parse_input(load_day(__day__))
# data = parse_input('''2-4,6-8
# 2-3,4-5
# 5-7,7-9
# 2-8,3-7
# 6-6,4-6
# 2-6,4-8'''.split('\n'))

if __name__ == '__main__':
    print(f'Day {__day__}: {part1(data)}, {part2(data)}')

File: 2016/src/bin/day04.rs
use std::{collections::HashMap, cmp::Ordering};
use aoc::{read_day, Solution};


fn main() {
    println!("Solution for day 4: {}", day4());
}

pub fn day4() -> Solution<u32> {
    let codes = parse_input(read_day(4));
    Solution::new(part1(&codes), part2(&codes))
}

fn parse_input(input: String) -> Vec<Code> {
    input
        .split("\n")
        .filter(|row| { !row.is_empty() })
        .map(|raw| {Code::from(raw)})
        .collect()
}

fn part1(codes: &Vec<Code>) -> u32 {
    codes.iter()
        .filter(|code| {code.is_valid()})
        .map(|c| c.sector)
        .sum()
}

fn part2(codes: &Vec<Code>) -> u32 {
    let target_room = "northpoleobjectstorage";
    let code: Vec<&Code> = codes.iter()
        .filter(|c| c.is_valid())
        .filter(|c| c.decypher() == target_room)
        .collect();
    code[0].sector()
}


struct Code {
    chars: Vec<char>,
    sector: u32,
    checksum: String,
}

impl Code {
    fn from(raw_code: &str) -> Self {
        const SPLIT_PATTERNS: [char; 2] = ['[', '-'];
        let mut tokens = raw_code
            .trim_end_matches(']')
            .split(SPLIT_PATTERNS)
            .collect::<Vec<&str>>();
        // assert_eq!(tokens.len(), 2);
        Code{
            checksum: tokens.pop()
                .expect("malformed checksum").to_string(),
            sector: tokens.pop()
                .expect("malformed sector").parse().expect("parsing error"),
            chars: tokens.iter().flat_map(|t| {t.chars()}).collect(),
        }
    }

    fn compute_checksum(&self) -> String {
        let counts: HashMap<char, u32> = self.chars.iter()
            .fold(HashMap::new(), |mut map, c| {
                *map.entry(*c).or_insert(0) += 1;
                map
            });

        let mut freqs: Vec<(char, u32)> = counts
            .into_iter()
            .collect::<Vec<(char, u32)>>();
        
        freqs.sort_by(|a, b| Code::compare_chars(*a, *b));
        freqs.iter().map(|(char, _)| char).take(5).collect()
    }

    fn compare_chars(token1: (char, u32), token2: (char, u32)) -> Ordering {
        match token2.1.cmp(&token1.1) {
            Ordering::Equal => token1.0.cmp(&token2.0),
            other => other,
        }
    }

    fn is_valid(&self) -> bool {
        self.checksum == self.compute_checksum()
    }

    fn decypher(&self) -> String {
        self.chars.iter()
            .map(|c| self.decypher_char(*c))
            .collect::<String>()
    }

    fn decypher_char(&self, c: char) -> char {
        ((c as u32 + self.sector - 97) % (123 - 97) + 97) as u8 as char
    }

    fn sector(&self) -> u32 {
        self.sector
    }
}
